distributed/shuffle/arrow.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    try:
+        import pyarrow as pa
+    except ImportError:
+        raise ImportError("PyArrow is needed for fast shuffling")
+
+
+def dump_batch(batch, file, schema=None) -> None:
+    """
+    Dump a batch to file, if we're the first, also write the schema
+
+    See Also
+    --------
+    load_arrow
+    """
+    if file.tell() == 0:
+        file.write(schema.serialize())
+    file.write(batch)
+
+
+def load_arrow(file) -> pa.Table:
+    """Load batched data written to file back out into a table again
+
+    Example
+    -------
+    >>> t = pa.Table.from_pandas(df)  # doctest: +SKIP
+    >>> with open("myfile", mode="wb") as f:  # doctest: +SKIP
+    ...     for batch in t.to_batches():  # doctest: +SKIP
+    ...         dump_batch(batch, f, schema=t.schema)  # doctest: +SKIP
+
+    >>> with open("myfile", mode="rb") as f:  # doctest: +SKIP
+    ...     t = load_arrow(f)  # doctest: +SKIP
+
+    See Also
+    --------
+    dump_batch
+    """
+    import pyarrow as pa
+
+    try:
+        sr = pa.RecordBatchStreamReader(file)
+        return sr.read_all()
+    except Exception:
+        raise EOFError
+
+
+def list_of_buffers_to_table(data: list[pa.Buffer], schema: pa.Schema) -> pa.Table:
+    """Convert a list of arrow buffers and a schema to an Arrow Table"""
+    import io
+
+    import pyarrow as pa
+
+    bio = io.BytesIO()
+    bio.write(schema.serialize())
+    for batch in data:
+        bio.write(batch)
```

Comment:
You and me both.  Yeah, Arrow is great if you use all Arrow primitives, but if you want to compose it with other things it's not that great because of all of the views and the lack of a deserialize function.  If we stick with Arrow (maybe?) then we should upstream a bunch of issues for this use case so that it gets cleaner in future releases.

I actually did a bit of performance tuning here.  Memory copies aren't significant yet, but I've reduced them to the extent that I can.